aptos-move/mvhashmap/src/unsync_map.rs

```diff
@@ -74,78 +83,107 @@ impl<
-                    .flat_map(|(t, v)| v.bytes().map(|bytes| (t, bytes))),
+                    .flat_map(|(t, v)| v.bytes_len().map(|s| (t, s))),
             )?),
             None => GroupReadResult::Uninitialized,
         })
     }
 
-    pub fn get_value_from_group(&self, group_key: &K, value_tag: &T) -> GroupReadResult {
+    pub fn fetch_group_tagged_data(
+        &self,
+        group_key: &K,
+        value_tag: &T,
+    ) -> Result<ValueWithLayout<V>, UnsyncGroupError> {
         self.group_cache.borrow().get(group_key).map_or(
-            GroupReadResult::Uninitialized,
+            Err(UnsyncGroupError::Uninitialized),
             |group_map| {
-                GroupReadResult::Value(
-                    group_map.get(value_tag).and_then(|v| v.extract_raw_bytes()),
-                    // TODO[agg_v2]: support layouts.
-                    None,
-                )
+                group_map
+                    .borrow()
+                    .get(value_tag)
+                    .cloned()
+                    .ok_or(UnsyncGroupError::TagNotFound)
             },
         )
     }
 
     /// Contains the latest group ops for the given group key.
-    pub fn finalize_group(&self, group_key: &K) -> Vec<(T, Arc<V>)> {
+    pub fn finalize_group(&self, group_key: &K) -> Vec<(T, ValueWithLayout<V>)> {
         self.group_cache
             .borrow()
             .get(group_key)
             .expect("Resource group must be cached")
+            .borrow()
             .clone()
             .into_iter()
             .collect()
     }
 
-    pub fn insert_group_op(&self, group_key: &K, value_tag: T, v: V) -> anyhow::Result<()> {
+    pub fn insert_group_op(
+        &self,
+        group_key: &K,
+        value_tag: T,
+        v: V,
+        maybe_layout: Option<Arc<MoveTypeLayout>>,
+    ) -> anyhow::Result<()> {
+        use anyhow::bail;
+        use aptos_types::write_set::WriteOpKind::*;
         use std::collections::hash_map::Entry::*;
-        use WriteOpKind::*;
-
         match (
             self.group_cache
                 .borrow_mut()
                 .get_mut(group_key)
                 .expect("Resource group must be cached")
+                .borrow_mut()
                 .entry(value_tag.clone()),
             v.write_op_kind(),
         ) {
             (Occupied(entry), Deletion) => {
                 entry.remove();
             },
             (Occupied(mut entry), Modification) => {
-                entry.insert(Arc::new(v));
+                entry.insert(ValueWithLayout::Exchanged(Arc::new(v), maybe_layout));
             },
             (Vacant(entry), Creation) => {
-                entry.insert(Arc::new(v));
+                entry.insert(ValueWithLayout::Exchanged(Arc::new(v), maybe_layout));
             },
-            (_, _) => {
+            (l, r) => {
+                println!("WriteOp kind {:?} not consistent with previous value at tag {:?}. l: {:?}, r: {:?}", v.write_op_kind(), value_tag, l, r);
```

Comment:
bail might not get printed? (it depends on the caller)

similarly in code_invariant_error we both log and raise PanicError

thought this can be error! instead of println!